Mathlib/Combinatorics/Quiver/Path/Vertices.lean

```diff
@@ -0,0 +1,262 @@
+/-
+Copyright (c) 2025 Matteo Cipollina. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Matteo Cipollina
+-/
+
+import Mathlib.Algebra.Order.Group.Nat
+import Mathlib.Algebra.Order.Sub.Basic
+import Mathlib.Combinatorics.Quiver.Path
+import Mathlib.Data.Finset.Dedup
+
+/-!
+# Path Vertices
+
+This file provides lemmas for reasoning about the vertices of a path.
+-/
+
+namespace Quiver.Path
+
+open List
+
+variable {V : Type*} [Quiver V]
+
+/-- The end vertex of a path. A path `p : Path a b` has `p.end = b`. -/
+def «end» {a : V} : ∀ {b : V}, Path a b → V
+  | b, _ => b
+
+@[simp]
+lemma end_cons {a b c : V} (p : Path a b) (e : b ⟶ c) : «end» (p.cons e) = c := rfl
+
+/-- The list of vertices in a path, including the start and end vertices. -/
+def vertices {a : V} : ∀ {b : V}, Path a b → List V
+  | _, nil => [a]
+  | _, cons p e => (p.vertices).concat («end» (p.cons e))
+
+/-- The set of vertices in a path. -/
+def vertices_set {a b : V} (p : Path a b) : Set V :=
+  {v | v ∈ p.vertices}
+
+/-- The finset of vertices in a path. -/
+def vertices_finset [DecidableEq V] {a b : V} (p : Path a b) : Finset V :=
+  p.vertices.toFinset
+
+/-- The finset of vertices in a path, excluding the final vertex. -/
+def init_finset [DecidableEq V] {a b : V} (p : Path a b) : Finset V :=
+  p.vertices.dropLast.toFinset
+
+@[simp]
+lemma mem_vertices_set_iff {a b : V} (p : Path a b) (v : V) :
+    v ∈ p.vertices_set ↔ v ∈ p.vertices := Iff.rfl
+
+@[simp]
+lemma mem_vertices_finset_iff [DecidableEq V] {a b : V} (p : Path a b) {x : V} :
+    x ∈ p.vertices_finset ↔ x ∈ p.vertices :=
+  List.mem_toFinset
+
+@[simp]
+lemma mem_init_finset_iff [DecidableEq V] {a b : V} (p : Path a b) {x : V} :
+    x ∈ p.init_finset ↔ x ∈ p.vertices.dropLast :=
+  List.mem_toFinset
+
+lemma vertices_nil (a : V) : (nil : Path a a).vertices = [a] := rfl
+
+@[simp]
+lemma vertices_cons {a b c : V} (p : Path a b) (e : b ⟶ c) :
+  (p.cons e).vertices = p.vertices.concat c := rfl
+
+-- Vertices of a path are always non-empty
+lemma vertices_nonempty' {V : Type*} [Quiver V] {a b : V} (p : Path a b) :
+    p.vertices.length > 0 := by
```

Comment:
```suggestion
lemma length_vertices_pos {a b : V} (p : Path a b) :
    0 < p.vertices.length := by
```

Reply:
Also if we move this lemma below `vertices_length` then the proof can be just `simp`.